fragment.go

```diff
@@ -1454,88 +1465,79 @@ func (f *fragment) bulkImport(rowIDs, columnIDs []uint64, options *ImportOptions
-		// order to maintain parity with the "normal" bulkimport path, we want
-		// to recalculate it at the end of the import
-		f.cache.Recalculate()
-		return nil
-	} // end "small update" path - after this is the real bulk import path
-
-	// Create a temporary bitmap which will be populated by rowIDs and columnIDs
-	// and then merged into the existing fragment's bitmap.
-	localBitmap := roaring.NewBitmap()
+		smallWrite = true
+		// If the number of operations is small, we'll write directly to local storage
+		localBitmap = f.storage
+	} else {
+		// Create a temporary bitmap which will be populated by rowIDs and columnIDs
+		// and then merged into the existing fragment's bitmap.
+		localBitmap = roaring.NewBitmap()
 
-	// Disconnect op writer so we don't append updates.
-	localBitmap.OpWriter = nil
+		// Disconnect op writer so we don't append updates.
+		localBitmap.OpWriter = nil
+	}
 
-	// rowSet maintains the set of rowIDs present in this import.
-	// It allows the cache to be updated once per row, instead of once
-	// per bit.
+	// rowSet maintains the set of rowIDs present in this import. It allows the
+	// cache to be updated once per row, instead of once per bit. TODO: consider
+	// sorting by rowID/columnID first and avoiding the map allocation here. (we
+	// could reuse rowIDs to store the list of unique row IDs)
 	rowSet := make(map[uint64]struct{})
 	lastRowID := uint64(0)
 
-	// Process every bit by writing to a local bitmap,
-	// to be merged with fragment storage next.
-	for i := range rowIDs {
+	// replace columnIDs with calculated positions to avoid allocation.
+	for i := 0; i < len(columnIDs); i++ {
 		rowID, columnID := rowIDs[i], columnIDs[i]
-
-		// Determine the position of the bit in the storage.
 		pos, err := f.pos(rowID, columnID)
 		if err != nil {
 			return err
 		}
-
-		// Write to local storage.
-		_, err = localBitmap.Add(pos)
-		if err != nil {
-			return err
-		}
-
-		// Reduce the StatsD rate for high volume stats
-		f.stats.Count("ImportBit", 1, 0.0001)
+		columnIDs[i] = pos
 
 		// Add row to rowSet.
 		if i == 0 || rowID != lastRowID {
 			lastRowID = rowID
 			rowSet[rowID] = struct{}{}
 		}
 	}
+	positions := columnIDs
+	var changedN int
+	if options.Clear {
+		changedN, err = localBitmap.RemoveN(positions...) // TODO benchmark AddN behavior with sorted/unsorted positions
+	} else {
+		changedN, err = localBitmap.AddN(positions...) // TODO benchmark RemoveN behavior with sorted/unsorted positions
+	}
+	if err != nil {
+		return errors.Wrap(err, "adding positions")
+	}
+	f.stats.Count("ImportBit", int64(changedN), 1)
+	f.opN += changedN
 
-	f.mu.Lock()
-	defer f.mu.Unlock()
-
-	// Merge localBitmap into fragment's existing data.
 	var results *roaring.Bitmap
-	if options.Clear {
-		if f.storage.Count() > 0 {
-			results = f.storage.Difference(localBitmap)
+	if !smallWrite {
+		// Merge localBitmap into fragment's existing data.
+		if options.Clear {
+			if f.storage.Count() > 0 {
```

Comment:
I wonder if we should have a short-circuit `Any()` which we can use for cases where we just care whether the count is 0 or not-0.